go/mysql/server.go

```diff
@@ -343,65 +343,74 @@ func (l *Listener) handle(conn net.Conn, connectionID uint32, acceptTime time.Ti
+				// sendFinished is set if the response should just be an OK packet.
+				sendFinished := false
+				err := l.handler.ComQuery(c, query, func(qr *sqltypes.Result) error {
+					if sendFinished {
+						// Failsafe: Unreachable if server is well-behaved.
+						return io.EOF
+					}
 
-				if !fieldSent {
-					fieldSent = true
+					if !fieldSent {
+						fieldSent = true
+
+						if len(qr.Fields) == 0 {
+							sendFinished = true
+							// We should not send any more packets after this.
+							return c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
+						}
+						if err := c.writeFields(qr); err != nil {
+							return err
+						}
+					}
 
-					if len(qr.Fields) == 0 {
-						sendFinished = true
-						// We should not send any more packets after this.
-						return c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
+					return c.writeRows(qr)
+				})
+
+				// If no field was sent, we expect an error.
+				if !fieldSent {
+					// This is just a failsafe. Should never happen.
+					if err == nil || err == io.EOF {
+						err = NewSQLErrorFromError(errors.New("unexpected: query ended without no results and no error"))
 					}
-					if err := c.writeFields(qr); err != nil {
-						return err
+					if werr := c.writeErrorPacketFromError(err); werr != nil {
+						// If we can't even write the error, we're done.
+						log.Errorf("Error writing query error to %s: %v", c, werr)
+						return false
 					}
+					return true
 				}
 
-				return c.writeRows(qr)
-			})
-
-			// If no field was sent, we expect an error.
-			if !fieldSent {
-				// This is just a failsafe. Should never happen.
-				if err == nil || err == io.EOF {
-					err = NewSQLErrorFromError(errors.New("unexpected: query ended without no results and no error"))
+				if err != nil {
+					// We can't send an error in the middle of a stream.
+					// All we can do is abort the send, which will cause a 2013.
+					log.Errorf("Error in the middle of a stream to %s: %v", c, err)
+					return false
 				}
-				if werr := c.writeErrorPacketFromError(err); werr != nil {
-					// If we can't even write the error, we're done.
-					log.Errorf("Error writing query error to %s: %v", c, werr)
-					return
+
+				// Send the end packet only sendFinished is false (results were streamed).
+				if !sendFinished {
+					if err := c.writeEndResult(false); err != nil {
+						log.Errorf("Error writing result to %s: %v", c, err)
+						return false
+					}
 				}
-				continue
-			}
 
-			if err != nil {
-				// We can't send an error in the middle of a stream.
-				// All we can do is abort the send, which will cause a 2013.
-				log.Errorf("Error in the middle of a stream to %s: %v", c, err)
+				timings.Record(queryTimingKey, queryStart)
+				return true
+			}()
+			if !success {
 				return
 			}
 
-			// Send the end packet only sendFinished is false (results were streamed).
-			if !sendFinished {
-				if err := c.writeEndResult(false); err != nil {
```

Comment:
there's a subtle behavior change here that i don't understand enough to know if it's a problem.

before -- the flush() was inline with writeEndResult, so we'd exit out of the handle loop if the flush() failed.
now -- flush() is in a non-blocking, non caught defer(), so it fails silently and we continue on our merry way

Reply:
i think if you just pull the startBuffering() and defer flush() out of the sub-function, then you'll be in good shape.

Reply:
I did that and also removed the function wrapper because it became unnecessary. Code is not as elegant, but will do for now.